### PR TITLE
[kernel] add kernels for llama inference

### PR DESCRIPTION
**Pull request by tiandiao123**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [ ] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [ ] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

## 📝 What does this PR do?
This PR is used to add llama useful cuda kernels for inference. 

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [x] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [x] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [x] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: requirements/requirements.txt [hunk 1]
```diff
@@ -10,4 +10,5 @@ contexttimer
 ninja
 torch>=1.12
 safetensors
+flash_attn==2.0.5
```
Comment (ver217) [inline requirements/requirements.txt:13]: Please move this to requirements-test

Comment (tiandiao123) [reply requirements/requirements.txt]: fixed.

**Review by ver217 — APPROVED** [review]

**Review by kurisusnowdeng — COMMENTED** [review]
Any differece from the original codes? For example, I quickly checked `pos_encoding` & `layernorm` kernels but they seem exactly the same as the codes in vllm's repository. 
If this is just integrating some external codes into our project, I recommend just to add the external tool as a dependency and import the necessary API for our needs.

**Comment by tiandiao123** [comment]
> Any differece from the original codes? For example, I quickly checked `pos_encoding` & `layernorm` kernels but they seem exactly the same as the codes in vllm's repository. If this is just integrating some external codes into our project, I recommend just to add the external tool as a dependency and import the necessary API for our needs.
There is some minor changes. I only integrate useful parts which will be integrated into our llama-inference attention over here.

**Comment by tiandiao123** [comment]
> Any differece from the original codes? For example, I quickly checked `pos_encoding` & `layernorm` kernels but they seem exactly the same as the codes in vllm's repository. If this is just integrating some external codes into our project, I recommend just to add the external tool as a dependency and import the necessary API for our needs.

There is some minor changes. I only integrate useful parts which will be integrated into our llama-inference attention over here.

**Comment by tiandiao123** [comment]
Adding llama shardformer demo, try to directly import third-party useful ops @kurisusnowdeng

**Comment by tiandiao123** [comment]
temporarily closed this.

**Pull request closed without merging**